### PR TITLE
✨ DateTimeLib

### PR DESCRIPTION
**Pull request by Vectorized**
## Description

✨

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge snapshot`?
- [x] Ran `npm run lint`?
- [x] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

**Comment by Vectorized** [comment]
@Atarpara @RogerPodacter 

- Removed `endOfWeek`. Cuz sunday is ambiguious. Some treat Sunday as start of week, some treat Sunday as end of week. Also, it is ambiguious whether the timestamp is for `00:00` or `23:59`, for the name `endOfWeek`.

- Renamed `getStartOfWeek` to `mondayTimestamp`. Monday is not an ambiguious day of week.

- Removed `getNextWeekday`. Can be calculated via `mondayTimestamp` and `dayOfWeek`.

- Restricted max supported year to `type(uint32).max`. By the time, we will already have been either extinct, or an interstellar species, or the Earth's motion would have drastically changed. A smaller supported range will mean smaller bytecode needed to validate the dates.

- Renamed functions `yFromX` to `xToY`. For consistency reasons.

**Comment by Vectorized** [comment]
@Atarpara I see you used a lot of neat tricks hehe.

Let this marinate for a while. I think the main focus now is to make sure it is absolutely correct.

**Comment by Atarpara** [comment]
@Vectorized check the testcase i covered most of the edge case but you changed the year limit so let me check.

**Comment by Atarpara** [comment]
@Vectorized checkout #215 PR. Test are fixed.

**Comment by Vectorized** [comment]
@Atarpara Is the algorithm in https://howardhinnant.github.io/date_algorithms.html is faster than the one used by https://github.com/bokkypoobah/BokkyPooBahsDateTimeLibrary/blob/master/contracts/BokkyPooBahsDateTimeLibrary.sol ?

**Comment by Vectorized** [comment]
@Atarpara Nice trick `add(shr(11, add(mul(62719, mod(add(month, 9), 12)), 769)), day)`.

**Comment by Vectorized** [comment]
I've added more tests, if you want, can see if `dateToEpochDay` and `epochDayToDate` can be optimized further.

File: src/utils/DateTimeLib.sol [hunk 1]
```diff
@@ -21,7 +21,9 @@ library DateTimeLib {
     /*                         CONSTANTS                          */
     /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
 
-    // Weekdays are 1-indexed for a traditional rustic feel.
+    // "And on the seventh day God finished his work that he had done,
```
Comment (RogerPodacter) [inline src/utils/DateTimeLib.sol:39]: LOL!
